src/corehost/corehost.cpp

```diff
@@ -226,25 +185,38 @@ int run(const int argc, const pal::char_t* argv[])
         return StatusCode::CoreHostCurExeFindFailure;
     }
 
-    pal::string_t app_root;
     pal::string_t app_path;
+    pal::string_t app_root;
     bool requires_v2_hostfxr_interface = false;
 
 #if FEATURE_APPHOST
-    pal::string_t app_dll_name;
-    if (!is_exe_enabled_for_execution(host_path, &app_dll_name))
+    pal::string_t embedded_app_name;
+    if (!is_exe_enabled_for_execution(host_path, &embedded_app_name))
     {
         trace::error(_X("A fatal error was encountered. This executable was not bound to load a managed DLL."));
         return StatusCode::AppHostExeNotBoundFailure;
     }
 
-    if (!resolve_app_root(host_path, &app_root, &requires_v2_hostfxr_interface))
+    if (_X('/') != DIR_SEPARATOR)
     {
+        replace_char(&embedded_app_name, _X('/'), DIR_SEPARATOR);
+    }
+
+    auto pos_path_char = embedded_app_name.find(DIR_SEPARATOR);
+    if (pos_path_char != pal::string_t::npos)
+    {
+        requires_v2_hostfxr_interface = true;
+    }
+
+    app_path.assign(get_directory(host_path));
+    append_path(&app_path, embedded_app_name.c_str());
+    if (!pal::realpath(&app_path))
+    {
+        trace::error(_X("The application to execute does not exist: '%s'."), app_path.c_str());
```

Comment:
Review note: this is the same message that already exists in hostfxr; we now also check for a valid file in apphost since we need a valid path to look for a local hostfxr.dll.